### PR TITLE
Add RSpec & raising errors when files are inaccessible

### PR DESCRIPTION
**Pull request by sjanusz-r7**
This PR is a pre-requisite for https://github.com/rapid7/metasploit-framework/issues/16473

This PR uses the recent `manifest` and `manifest.uuid` files created by a Rake task to keep track of all the bundled files and their `SHA3-256` hashes.
This allows us to see if any of the files present in the `meterpreter` directory have been deleted or that the user does not have permission to read and to notify the user.

The implemented API allows us to call:
```ruby
file_warnings = ::MetasploitPayloads.file_warnings
file_warnings.each { |file| puts "#{file[:path]} : #{file[:reason]} } 
```

resulting in an example output of:
```ruby
[
    {:path=>"/Users/sjanusz/Programming/metasploit-payloads/gem/data/meterpreter/ext_server_stdapi.x64.dll", :reason=>:not_readable},
    {:path=>"meterpreter/metsrv.x64.dll", :reason=>:not_found},
    {:path=>"meterpreter/screenshot.x64.debug.dll", :reason=>:not_found},
    {:path=>"meterpreter/meterpreter.php", :reason=>:hash_mismatch}
]
```

To make the errors more end-user friendly, map the `reason` to a string, potentially resulting in:
```ruby
/Users/sjanusz/Programming/metasploit-payloads/gem/data/meterpreter/ext_server_stdapi.x64.dll : No read permissions
meterpreter/metsrv.x64.dll : File not found
meterpreter/screenshot.x64.debug.dll : File not found
meterpreter/meterpreter.php : Hash mismatch
```

This PR also adds RSpec and tests which can be executed in the `gem` directory with:
```ruby
bundle exec 'rspec ./spec/spec_helper.rb'
```

### CODE REVIEW

File: gem/lib/metasploit-payloads/version.rb [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # -*- coding:binary -*-
 module MetasploitPayloads
-  VERSION = '2.0.150'
+  VERSION = '2.0.151'
```
Comment (dwelch-r7) [inline gem/lib/metasploit-payloads/version.rb:3]: You can revert this and rebase on master

File: gem/lib/metasploit-payloads/file_list.rb
Comment (dwelch-r7) [inline gem/lib/metasploit-payloads/file_list.rb:1]: Let's circle back to this and discuss a better UX for when new extensions etc. are added

File: gem/lib/metasploit-payloads.rb [hunk 2]
```diff
@@ -10,48 +12,126 @@ module MetasploitPayloads
...
+  #
+  # @return [Array<Hash<String, Symbol>>] An array of filenames with warnings. Provides a file name and reason.
+  #  Empty if all needed Meterpreter files exist and have the correct hash.
+  def self.file_warnings
+    file_warnings = []
+
+    begin
+      manifest_contents = ::File.binread(manifest_path)
+    rescue ::Errno::ENOENT => _e
+      return [{ path: 'manifest', reason: :not_found }]
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:24]: Can we update the paths here to always be absolute, as it's hard from a caller's perspective to know which type of value this function is returning

File: gem/lib/metasploit-payloads.rb [hunk 3]
```diff
@@ -10,48 +12,126 @@ module MetasploitPayloads
...
+    rescue ::Errno::EACCES => _e
+      return [{path: 'manifest', reason: :not_readable}]
+    end
+
+    begin
+      manifest_uuid_contents = ::File.binread(manifest_uuid_path)
+    rescue ::Errno::ENOENT => _e
+      file_warnings.append({path: 'manifest.uuid', reason: :not_found})
+    rescue ::Errno::EACCES => _e
+      file_warnings.append({path: 'manifest.uuid', reason: :not_readable})
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:34]:
```suggestion
    rescue => e
      file_warnings.append({ path: 'manifest.uuid', error: e })
```

File: gem/lib/metasploit-payloads.rb [hunk 4]
```diff
@@ -10,48 +12,126 @@ module MetasploitPayloads
...
+        out_path = self.path(filename)
+        bundled_file = out_path.start_with?(data_directory)
+        if bundled_file
+          file_hash_match = (::OpenSSL::Digest.new(hash_type, read(filename)).to_s == hash)
+          file_warnings.append({path: filename, reason: :hash_mismatch}) unless file_hash_match
+        end
+      rescue ::MetasploitPayloads::NotFoundError => e
+        file_warnings.append({path: e.path, reason: :not_found})
+      rescue ::MetasploitPayloads::NotReadableError => e
+        file_warnings.append({path: e.path, reason: :not_readable})
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:60]:
```suggestion
      rescue ::MetasploitPayloads::NotFoundError, ::MetasploitPayloads::NotReadableError => e
        file_warnings.append({ path: e.path, error: e })
      rescue => e
         file_warnings.append({ path: filename_with_data, error: e })       
```

File: gem/lib/metasploit-payloads.rb [hunk 5]
```diff
@@ -10,48 +12,126 @@ module MetasploitPayloads
   METERPRETER_SUBFOLDER = 'meterpreter'
   USER_DATA_SUBFOLDER   = 'payloads'
 
+  #
+  # @return [Array<Hash<String, Symbol>>] An array of filenames with warnings. Provides a file name and reason.
+  #  Empty if all needed Meterpreter files exist and have the correct hash.
+  def self.file_warnings
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:18]: Maybe a better name? 🤔 

```suggestion
  def self.manifest_errors
```

Rationale being, the caller should decide what's a warning/error/etc

File: gem/lib/metasploit-payloads.rb [hunk 6]
```diff
@@ -10,48 +12,126 @@ module MetasploitPayloads
...
+    end
+
+    slice_length = './data/'.length
+
+    manifest_contents.each_line do |line|
+      filename_with_data, hash_type, hash = line.chomp.split(':')
+      # We need to remove the './data/' from the start of each filename as path() prepends the 'data' sub-folder.
+      filename = filename_with_data[slice_length..-1]
+      begin
+        out_path = self.path(filename)
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:51]: Maybe self.read here? 🤔

File: gem/lib/metasploit-payloads.rb [hunk 7]
```diff
@@ -10,48 +12,126 @@ module MetasploitPayloads
...
+        # rather than having MSF Console falling back to the files in the gem
+        raise ::MetasploitPayloads::NotReadableError, extra_path, caller if ::File.exist?(extra_path)
       end
     end
 
     return gem_path if ::File.readable? gem_path
 
-    nil
+    raise ::MetasploitPayloads::NotReadableError, gem_path, caller if ::File.exist?(gem_path)
   end
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:102]:
```ruby
    return gem_path if ::File.readable? gem_path
    raise ::MetasploitPayloads::NotReadableError, gem_path, caller if ::File.exist?(gem_path)
    
    nil
  end
```

File: gem/lib/metasploit-payloads.rb [hunk 8]
```diff
@@ -10,48 +12,126 @@ module MetasploitPayloads
...
 
+  #
+  # Get the path for the first readable path in the provided arguments.
+  # Start with the provided `extra_paths` then fall back to the `gem_path`.
+  #
+  # @param [String] gem_path a path to the gem
+  # @param [Array<String>] extra_paths a path to any extra paths that should be evaluated for local files before `gem_path`
+  # @raise [NotReadableError] if the user doesn't have read permissions for the currently-evaluated path
+  # @return [String] A readable path
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:81]:
```suggestion
  # @return [String,nil] A readable path or nil
```

File: gem/lib/metasploit-payloads.rb [hunk 9]
```diff
@@ -88,6 +168,14 @@ def self.list_meterpreter_extension_suffixes(extension_name=nil)
     list_meterpreter_dirs { |dir| meterpreter_enum_ext_suffixes(dir, extension_name) }
   end
 
+  def self.manifest_path
+    ::File.realpath(::File.join(::File.dirname(__FILE__), '..', 'manifest'))
+  end
+
+  def self.manifest_uuid_path
+    ::File.realpath(::File.join(::File.dirname(__FILE__), '..', 'manifest.uuid'))
+  end
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:177]: It'd be great if this was private, if it's easy enough to add it to the eigenclass beneath and the tests still pass - that works for me.

If not, we can circle back and make it private

File: gem/lib/metasploit-payloads/error.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+module MetasploitPayloads
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads/error.rb:3]: Rubocop 👮‍♂️

File: gem/spec/metasploit_payloads/metasploit_payloads_spec.rb [hunk 11]
```diff
@@ -0,0 +1,252 @@
+# frozen_string_literal: true
+
+require 'metasploit-payloads'
+
+RSpec.describe ::MetasploitPayloads do
```
Comment (adfoster-r7) [inline gem/spec/metasploit_payloads/metasploit_payloads_spec.rb:5]: rubocop 👮‍♂️

File: gem/spec/metasploit_payloads/metasploit_payloads_spec.rb [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+        expect { subject.send(:readable_path, @sample_file) }.to raise_error(::MetasploitPayloads::NotReadableError)
+      end
+    end
+
+    context 'when the path does not exist' do
+      it 'returns nil' do
+        allow(::File).to receive(:exist?).with(@sample_file).and_return(false)
+        allow(::File).to receive(:readable?).with(@sample_file).and_return(false)
+
+        expect(subject.send(:readable_path, @sample_file)).to eq(nil)
```
Comment (adfoster-r7) [inline gem/spec/metasploit_payloads/metasploit_payloads_spec.rb:50]:
```suggestion
        expect(subject.readable_path, @sample_file).to eq(nil)
```

File: gem/spec/metasploit_payloads/metasploit_payloads_spec.rb [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+    end
+
+    context 'when the path is not readable' do
+      it 'raises a ::MetasploitPayloads::NotReadableError' do
+        allow(::File).to receive(:exist?).with(@sample_file).and_return(true)
+        allow(::File).to receive(:readable?).with(@sample_file).and_return(false)
+
+        expect { subject.send(:readable_path, @sample_file) }.to raise_error(::MetasploitPayloads::NotReadableError)
+      end
+    end
```
Comment (adfoster-r7) [inline gem/spec/metasploit_payloads/metasploit_payloads_spec.rb:43]: https://www.betterspecs.org/#let

```suggestion
    let(:sample_file) { 'meterpreter/meterpreter.py' }

    before :each do
      allow(::File).to receive(:exist?).and_call_original
      allow(::File).to receive(:readable?).and_call_original
    end

    context 'when the path is not readable' do
      it 'raises a ::MetasploitPayloads::NotReadableError' do
        allow(::File).to receive(:exist?).with(sample_file).and_return(true)
        allow(::File).to receive(:readable?).with(sample_file).and_return(false)
        expect { subject.readable_path(sample_file) }.to raise_error(::MetasploitPayloads::NotReadableError)
      end
    end
```

File: gem/spec/metasploit_payloads/metasploit_payloads_spec.rb [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+        end
+      end
+
+      context 'cannot be read' do
+        it 'includes the correct error' do
+          path = ::MetasploitPayloads.manifest_uuid_path
+          allow(::File).to receive(:binread).with(path).and_raise(::Errno::EACCES.new(path))
+
+          expect(subject.send(:file_warnings)).to include({path: 'manifest.uuid', reason: :not_readable})
+        end
```
Comment (adfoster-r7) [inline gem/spec/metasploit_payloads/metasploit_payloads_spec.rb:217]:
```suggestion
```

File: gem/spec/metasploit_payloads/metasploit_payloads_spec.rb [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+      end
+    end
+
+    context 'when the manifest.uuid file' do
+      context 'does not exist' do
+        it 'includes the correct error' do
+          path = ::MetasploitPayloads.manifest_uuid_path
+          allow(::File).to receive(:binread).with(path).and_raise(::Errno::ENOENT.new(path))
+
+          expect(subject.send(:file_warnings)).to include({path: 'manifest.uuid', reason: :not_found})
```
Comment (adfoster-r7) [inline gem/spec/metasploit_payloads/metasploit_payloads_spec.rb:207]:
```suggestion
          error = 
          allow(::File).to receive(:binread).with(path).and_raise(error)

          expect(subject.send(:file_warnings)).to include({path: 'manifest.uuid', error: error})
```

File: gem/lib/metasploit-payloads.rb [hunk 16]
```diff
@@ -10,48 +12,127 @@ module MetasploitPayloads
   METERPRETER_SUBFOLDER = 'meterpreter'
   USER_DATA_SUBFOLDER   = 'payloads'
 
+  #
+  # @return [Array<Hash<String, Symbol>>] An array of filenames with warnings. Provides a file name and reason.
```
Comment (adfoster-r7) [inline gem/lib/metasploit-payloads.rb:16]:
```suggestion
  # @return [Array<Hash<String, Symbol>>] An array of filenames with warnings. Provides a file name and error.
```

**Pull request merged**